### PR TITLE
Add Container Insights.

### PR DESCRIPTION
**Pull request by awsimaya**
Adding Container Insights to the workshop. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by awsimaya** [comment]
Made changes per Brent's suggestion

File: content/container_insights/setup.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+weight: 1
+---
+
+There are 2 components that work together to collect Metrics, Logs data for Container Insights in EKS
+
+1. CloudWatch agent
+2. FluentD
+
+#### Application setup
+In this section we will be setting up Container Insights for the ECS Demo app that you created in the **Deploy the Example Microservices** section earlier. If you haven't done that yet, go ahead and deploy the application and come back here again to proceed with Container Insights.
```
Comment (dalbhanj) [inline content/container_insights/setup.md]: Link to **Deploy the Example Microservices** chapter

File: content/container_insights/setup.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+------------------------------------------------------------
+
+#### Setup IAM role
```
Comment (dalbhanj) [inline content/container_insights/setup.md]: This entire step can be automated using scripts. You can see here for inspiration https://eksworkshop.com/scaling/deploy_ca/

File: content/container_insights/setup.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+If you are using Linux or macOS, run the following command to setup Container Insights on your EKS cluster. Make sure you replace **<CLUSTER_NAME>** and **<REGION_NAME>** with your EKS cluster name and AWS region the cluster is in.
+
+`curl https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/master/k8s-yaml-templates/quickstart/cwagent-fluentd-quickstart.yaml | sed "s/{{cluster_name}}/Cluster_Name/;s/{{region_name}}/Region/" | kubectl apply -f -`
+
+##### Windows users
+
+If you're using Windows, use the below command to setup Container Insights on your EKS cluster.Make sure you replace **<CLUSTER_NAME>** and **<REGION_NAME>** with your EKS cluster name and AWS region the cluster is in.
+
+`kubectl apply -f (New-Item -ItemType "File" -Name "containerinsights.yml" -Value ((Invoke-WebRequest -Uri https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/master/k8s-yaml-templates/quickstart/cwagent-fluentd-quickstart.yaml).Content -replace "{{cluster_name}}","<CLUSTER_NAME>" -replace "{{region_name}}","<REGION_NAME>"))`
```
Comment (dalbhanj) [inline content/container_insights/setup.md]: Look for other chapters for proper format (https://eksworkshop.com/intro_to_rbac/install_test_pods/)

File: content/container_insights/setup.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+------------------------------------------------------------
+
+#### Setup Container Insights
+
+##### Linux/macOS users
+
+If you are using Linux or macOS, run the following command to setup Container Insights on your EKS cluster. Make sure you replace **<CLUSTER_NAME>** and **<REGION_NAME>** with your EKS cluster name and AWS region the cluster is in.
+
+`curl https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/master/k8s-yaml-templates/quickstart/cwagent-fluentd-quickstart.yaml | sed "s/{{cluster_name}}/Cluster_Name/;s/{{region_name}}/Region/" | kubectl apply -f -`
```
Comment (dalbhanj) [inline content/container_insights/setup.md]: Look for other chapters for proper format (https://eksworkshop.com/intro_to_rbac/install_test_pods/)

File: content/container_insights/explore.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+* Navigate to Amazon CloudWatch home page. Select **Container Insights** from the drop down on the home page as shown below
+
+![Container Insights](/images/ContainerInsights1.png)
+
+* Ensure **EKS Clusters** is selected in the first dropdown and select the **eksworkshop-eksctl** cluster in the second dropdown. You will be able to see several build-in charts showing various cluster level metrics such as CPU Utilization, Memory Usage, Disk Usage, Network and so on in the default dashboard as shown below
+
+![Container Insights](/images/ContainerInsights2.png)
+
+* You can also view Performance Logs, Application Logs, Control Plane Logs ([when enabled](https://docs.aws.amazon.com/en_pv/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-control-plane-logging.html)), Data Plane Logs and Host Logs by simply selecting the cluster name and clicking on Actions dropdown as shown below
```
Comment (dalbhanj) [inline content/container_insights/explore.md]: I would add instructions into why is this interesting and provide steps to identify the importance of Control plane logging. I'd rather remove this section if we are not highlighting why is this important

File: content/container_insights/explore.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+![Container Insights](/images/ContainerInsights3.png)
+
+* You can also drill down into the cluster and see the metrics at the Pod level by simply selecting **EKS Pods** in the first drop down as shown below
+
+![Container Insights](/images/ContainerInsights5.png)
+
+* Likewise, you can also see insights at the Node level as well. You just have to select **EKS Nodes** in the first drop down and select a particular node in the second drop down to see metrics specific to that node.
```
Comment (dalbhanj) [inline content/container_insights/explore.md]: Add a picture that shows what users can expect

File: content/container_insights/explore.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+---
+title: "Explore Container Insights"
+chapter: false
+weight: 2
+---
+#### Validate Installation
+
+* Once the installation is complete, execute the following commands to ensure everything is setup correctly.
```
Comment (dalbhanj) [inline content/container_insights/explore.md]: remove higher level bullets

File: content/container_insights/explore.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    * Finally, let's check if the Pods are in Running status. Execute the following command to find that out.
+
+        `kubectl get pods -n amazon-cloudwatch`
+    
+        You should see 3 instances of **cloudwatch-agent** and **fluentd-cloudwatch** pods running.
+
+![Screenshot checking Container Insights installation](/images/ContainerInsights9.png) 
+
+#### Verify that the logs are streaming into CloudWatch
+* Login to AWS console and navigate to [Amazon CloudWatch Logs](https://console.aws.amazon.com/cloudwatch/home#logs:)
```
Comment (dalbhanj) [inline content/container_insights/explore.md]: remove higher level bullets

File: content/container_insights/setup.md (hunk 1 above)
Comment (awsimaya) [reply content/container_insights/setup.md]: Linked the chapter per suggestion

File: content/container_insights/setup.md (hunk 2 above)
Comment (awsimaya) [reply content/container_insights/setup.md]: Completed per suggestion

File: content/container_insights/setup.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+* In the list of policies, select the check box next to **CloudWatchAgentServerPolicy**. If necessary, use the search box to find this policy.
+
+* Choose **Attach policies**
+
+------------------------------------------------------------
+
+#### Setup Container Insights
+
+##### Linux/macOS users
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setup.md]: Uses tabs per suggestion

File: content/container_insights/setup.md (hunk 3 above)
Comment (awsimaya) [reply content/container_insights/setup.md]: Completed per suggestion

File: content/container_insights/setup.md (hunk 4 above)
Comment (awsimaya) [reply content/container_insights/setup.md]: Completed per suggestion

File: content/container_insights/explore.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+chapter: false
+weight: 2
+---
+#### Validate Installation
+
+* Once the installation is complete, execute the following commands to ensure everything is setup correctly.
+
+    * Execute the following command
+
+        `kubectl get namespaces`
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/explore.md]: Completed per suggestion

File: content/container_insights/explore.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+        You should be able to see a new namespace called **amazon-cloudwatch** listed amongst others
+
+    * Now execute the following command
+
+        `kubectl get daemonsets -n amazon-cloudwatch`
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/explore.md]: Completed per suggestion

File: content/container_insights/explore.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+* Login to AWS console and navigate to [Amazon CloudWatch Logs](https://console.aws.amazon.com/cloudwatch/home#logs:)
+* Type **/aws/containerinsights/eksworkshop-eksctl** in the filter texbox and press enter. You should be able to see 4 Log Groups as shown below each one of them containing Application, Dataplane, Host and Performance Log Streams respectively.
+
+![Container Insights](/images/ContainerInsights6.png)
+
+* Navigate to Amazon CloudWatch home page. Select **Container Insights** from the drop down on the home page as shown below
+
+![Container Insights](/images/ContainerInsights1.png)
+
+* Ensure **EKS Clusters** is selected in the first dropdown and select the **eksworkshop-eksctl** cluster in the second dropdown. You will be able to see several build-in charts showing various cluster level metrics such as CPU Utilization, Memory Usage, Disk Usage, Network and so on in the default dashboard as shown below
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/explore.md]: Fixed the typo

File: content/container_insights/explore.md (hunk 5 above)
Comment (awsimaya) [reply content/container_insights/explore.md]: Completed per suggestion

File: content/container_insights/explore.md (hunk 6 above)
Comment (awsimaya) [reply content/container_insights/explore.md]: Added picture

File: content/container_insights/setupalarm.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
+---
+title: "Setup CloudWatch Alarm"
+chapter: false
+weight: 3
+---
+#### Setup CloudWatch Alarm
+
+In this section, we will setup a CloudWatch alarm based on CPU Utilizaiton metric on a specific EKS worker node.
+
+* Select any particular node you want and select **View in metrics** from the **CPU Utilization** widget as shown below
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: Removed bullets

File: content/container_insights/setupalarm.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
+---
+title: "Setup CloudWatch Alarm"
+chapter: false
+weight: 3
+---
+#### Setup CloudWatch Alarm
+
+In this section, we will setup a CloudWatch alarm based on CPU Utilizaiton metric on a specific EKS worker node.
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: Completed per suggestion

File: content/container_insights/setupalarm.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+* Select any particular node you want and select **View in metrics** from the **CPU Utilization** widget as shown below
+
+![Container Insights](/images/ContainerInsights7.png)
+
+* Click on the bell icon (circled red in the screenshot below) to create the alarm
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: Removed bullets

File: content/container_insights/explore.md (hunk 7 above)
Comment (awsimaya) [reply content/container_insights/explore.md]: Removed bullets

File: content/container_insights/explore.md (hunk 8 above)
Comment (awsimaya) [reply content/container_insights/explore.md]: Removed bullets

File: content/container_insights/setupalarm.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+![Container Insights](/images/ContainerInsights8.png)
+
+* In the next screen, enter 70 in the threshold textbox and leave everything as is. Click **Next**
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: Fixed the typo

File: content/container_insights/setupalarm.md [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+![Container Insights](/images/ContainerInsights10.png)
+
+* In **Configure actions** screen, select an existing SNS topic if you already have one, or create a new topic by selecting **Create new topic**, give the SNS topic a name and provide an email address to receive notifications from the topic. Click **Next**
+![Container Insights](/images/ContainerInsights11.jpg)
+
+* In the **Add a description** screen, enter a unique name for the alarm and click **Next**
+* Click on **Create alarm** in the **Preview and create** screen. CloudWatch will send you a notification whenever the CPU utilization on the worker node goes beyond 70% for a period of 5 minutes
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: Added cleanup section

File: content/container_insights/cleanup.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+#### Delete Container Insights
+
+{{< tabs name="Setup Instructions" >}}
+{{{< tab name="Linux / macOS" include="cleanup_linuxmacos.md" />}}
+{{{< tab name="Windows" include="cleanup_windowsos.md" />}}
+{{< /tabs >}}
+
+{{% children showhidden="false" %}}
+
+#### Cleanup the EKS cluster
```
Comment (dalbhanj) [inline content/container_insights/cleanup.md]: Cleanup example microservices

File: content/container_insights/cleanup.md [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+Follow [these steps](../../deploy/cleanup) to clean up the EKS resoures
```
Comment (dalbhanj) [inline content/container_insights/cleanup.md]: to cleanup example microservices from your EKS cluster

File: content/container_insights/setup.md (hunk 1 above)
Comment (awsimaya) [reply content/container_insights/setup.md]: Changed per suggestion

File: content/container_insights/setup.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+weight: 1
+---
+
+There are 2 components that work together to collect Metrics, Logs data for Container Insights in EKS
+
+1. CloudWatch agent
+2. FluentD 
+
+#### Application setup
+In this section we will be setting up Container Insights for the ECS Demo app that you created in the [Deploy the Example Microservices](../../deploy) section earlier. If you haven't done that yet, go ahead and deploy the application and come back here again to proceed with Container Insights.
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setup.md]: Addressed per suggestion

File: content/container_insights/setup.md [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+#### Setup IAM policy
+
+In order for the EKS worker nodes to be able to send logs and metrics to CloudWatch, follow the steps below
+
+```
+test -n "$ROLE_NAME" && echo ROLE_NAME is "$ROLE_NAME" || echo ROLE_NAME is not set
+```
+If ROLE_NAME is not set, please review: [this](../../eksctl/test/)
+
+You could also go to the EC2 console and get the IAM Role the EKS nodes are assuming from there. If you do so, make sure you replace **$ROLE_NAME** in the below command to the name of the role you got from the EC2 console. **_Caution -  This is NOT the Role ARN, but the friendly role name which needs to be used here._**
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setup.md]: Changed per suggestion

File: content/container_insights/linuxmacos.md [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+disableToc: true
+hidden: true
+---
+
+##### Linux / macOS users
+
+Make sure you replace **<CLUSTER_NAME>** and **<REGION_NAME>** with your EKS cluster name and AWS region the cluster is in.
+
+```
+curl https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/master/k8s-yaml-templates/quickstart/cwagent-fluentd-quickstart.yaml | sed "s/{{cluster_name}}/Cluster_Name/;s/{{region_name}}/Region/" | kubectl apply -f -
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/linuxmacos.md]: Added export commands

File: content/container_insights/cleanup_linuxmacos.md [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+disableToc: true
+hidden: true
+---
+
+##### Linux / macOS users
+
+Make sure you replace **<CLUSTER_NAME>** and **<REGION_NAME>** with your EKS cluster name and AWS region the cluster is in.
+
+```
+curl https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/master/k8s-yaml-templates/quickstart/cwagent-fluentd-quickstart.yaml | sed "s/{{cluster_name}}/Cluster_Name/;s/{{region_name}}/Region/" | kubectl delete -f -
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/cleanup_linuxmacos.md]: Added export commands

File: content/container_insights/cleanup_windowsos.md [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+---
+
+##### Windows users
+
+Make sure you replace **<CLUSTER_NAME>** and **<REGION_NAME>** with your EKS cluster name and AWS region the cluster is in.
+
+
+```
+kubectl delete -f (New-Item -ItemType "File" -Name "containerinsights.yml" -Force -Value ((Invoke-WebRequest -Uri https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/master/k8s-yaml-templates/quickstart/cwagent-fluentd-quickstart.yaml).Content -replace "{{cluster_name}}","<CLUSTER_NAME>" -replace "{{region_name}}","<REGION_NAME>"))
+```
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/cleanup_windowsos.md]: Added PowerShell equivalent of export commands

File: content/container_insights/explore.md [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+kubectl get pods -n amazon-cloudwatch
+```
+    
+You should see 3 instances of **cloudwatch-agent** and **fluentd-cloudwatch** pods running.
+
+![Screenshot checking Container Insights installation](/images/ContainerInsights9.png) 
+
+#### Verify that the logs are streaming into CloudWatch
+Login to AWS console and navigate to [Amazon CloudWatch Logs](https://console.aws.amazon.com/cloudwatch/home#logs:)
+Type **/aws/containerinsights/eksworkshop-eksctl** in the filter texbox and press enter. You should be able to see 4 Log Groups as shown below each one of them containing Application, Dataplane, Host and Performance Log Streams respectively.
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/explore.md]: Done

File: content/container_insights/setupalarm.md [hunk 26]
```diff
@@ -0,0 +1,55 @@
+---
+title: "Setup CloudWatch Alarm"
+chapter: false
+weight: 3
+---
+#### Setup CloudWatch Alarm
+
+In this section, we will setup a CloudWatch alarm and send notification using SNS the number of Pod restarts goes greater than 1 in a 5 minute period.
+
+##### Steps
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: These are actually new ones. I checked the URL, it shows the the ones I have put.

File: content/container_insights/setupalarm.md [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+![Container Insights](/images/ContainerInsights10.png)
+
+* Review the details in the following screen and click **Create alarm**
+
+##### Trigger the alarm
+
+Get the name of the **ecsdemo-frontend** pod using the following command
+
+```
+kubectl get pods
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/setupalarm.md]: Done.

File: content/container_insights/logsinsights.md [hunk 28]
```diff
@@ -0,0 +1,57 @@
+---
+title: "CloudWatch Logs Insights"
+chapter: false
+weight: 5
+---
+#### Querying Logs from EKS
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/logsinsights.md]: Done

File: content/container_insights/cleanup.md (hunk 18 above)
Comment (awsimaya) [reply content/container_insights/cleanup.md]: Done

File: content/container_insights/cleanup.md (hunk 19 above)
Comment (awsimaya) [reply content/container_insights/cleanup.md]: Done

File: content/container_insights/explore.md [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+![Container Insights](/images/ContainerInsights3.png)
+
+You can also drill down into the cluster and see the metrics at the Pod level by simply selecting **EKS Pods** in the first drop down as shown below
+
+![Container Insights](/images/ContainerInsights5.png)
+
+Likewise, you can also see insights at the Node level as well. You just have to select **EKS Nodes** in the first drop down and select a particular node in the second drop down to see metrics specific to that node.
+
+![Container Insights](/images/ContainerInsights11.png)
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/explore.md]: Changed the pic. It was indeed wrong.

File: content/container_insights/explore.md [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+You should be able to see two daemonsets called **cloudwatch-agent** and **fluentd-cloudwatch** listed.
+Finally, let's check if the Pods are in Running status. Execute the following command to find that out.
+
+```
+kubectl get pods -n amazon-cloudwatch
+```
+    
+You should see 3 instances of **cloudwatch-agent** and **fluentd-cloudwatch** pods running.
+
+![Screenshot checking Container Insights installation](/images/ContainerInsights9.png)
```
[the remark this answers is not in the document]
Comment (awsimaya) [reply content/container_insights/explore.md]: I use Windows Terminal on my Windows 10 machine. I customized the background :) I really like it, please don't ask me to change 🙏  :)

File: content/container_insights/setupalarm.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
+---
+title: "Setup CloudWatch Alarm"
+chapter: false
+weight: 3
+---
+#### Setup CloudWatch Alarm
+
+In this section, we will setup a CloudWatch alarm and send notification using SNS the number of Pod restarts goes greater than 1 in a 5 minute period.
```
Comment (dalbhanj) [inline content/container_insights/setupalarm.md]: we will setup CloudWatch alarm and send notification using SNS when a Pod restarts more than once in a 5 minute period.

File: content/container_insights/setupalarm.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+![Container Insights](/images/ContainerInsights10.png)
+
+* Review the details in the following screen and click **Create alarm**
+
+##### Trigger the alarm
+
+Get the name of the **ecsdemo-frontend** pod using the following command
+
+```
+kubectl get pods
```
Comment (dalbhanj) [inline content/container_insights/setupalarm.md]: I'd just add both command and expected output. And Remove the table below, very nit picky but it helps with consistency
For ex, 
$ kubectl get pods --all-namespaces
NAMESPACE        NAME                                              READY   STATUS    RESTARTS   AGE
kube-system      aws-node-5pzjk                                    1/1     Running   0          3d22h
kube-system      aws-node-7dbgq                                    1/1     Running   0          3d22h
kube-system      aws-node-fwvd5                                    1/1     Running   0          3d22h

File: content/container_insights/logsinsights.md [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+
+You can also apply filters to the query as shown below
+
+```
+fields kubernetes.pod_name, kubernetes.host
+| filter kubernetes.pod_name like 'ecsdemo'
+| stats count_distinct(kubernetes.pod_name) as Number_of_Pods by kubernetes.host  
+```
+
+The above query will query will filter the pods with name containing the string **ecsdemo**. Your output should look similar to the one shown below.
```
Comment (dalbhanj) [inline content/container_insights/logsinsights.md]: The above query will filter the pods

**Pull request merged**